File: internal/middleware/cors.go
package middleware

import (
	"strings"

	"github.com/gin-gonic/gin"
)

func CORSMiddleware(origins []string) gin.HandlerFunc {
	originSet := make(map[string]bool, len(origins))
	for _, o := range origins {
		originSet[strings.TrimRight(o, "/")] = true
	}
	return func(c *gin.Context) {
		origin := c.Request.Header.Get("Origin")
		allowed := ""

		if originSet[origin] {
			allowed = origin
		} else if len(originSet) == 1 {
			for o := range originSet {
				allowed = o
			}
		} else if originSet["*"] {
			allowed = "*"
		}

		if allowed != "" {
			c.Header("Access-Control-Allow-Origin", allowed)
		}
		c.Header("Access-Control-Allow-Credentials", "true")
		c.Header("Access-Control-Allow-Headers", "Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token, Authorization, accept, origin, Cache-Control, X-Requested-With")
		c.Header("Access-Control-Allow-Methods", "POST, OPTIONS, GET, PUT, DELETE, PATCH")

		if c.Request.Method == "OPTIONS" {
			c.AbortWithStatus(204)
			return
		}

		c.Next()
	}
}

File: src/middleware/cors.ts
/**
 * Middleware CORS. Permite solicitudes desde orígenes configurados
 * y maneja preflight requests (OPTIONS).
 *
 * En producción refleja el Origin de la solicitud si es una URL válida,
 * permitiendo así cualquier frontend autorizado sin necesidad de
 * configurar cada origen manualmente.
 */
import type { Context, Next } from "hono";

/** Middleware que configura headers CORS reflejando el origen de la solicitud */
export async function corsMiddleware(c: Context, next: Next) {
  const origin = c.req.header("Origin");

  if (origin && (origin.startsWith("https://") || origin.startsWith("http://"))) {
    c.header("Access-Control-Allow-Origin", origin);
    c.header("Access-Control-Allow-Credentials", "true");
    c.header("Vary", "Origin");
  } else {
    c.header("Access-Control-Allow-Origin", "*");
  }

  c.header("Access-Control-Allow-Methods", "GET, POST, PUT, DELETE, PATCH, OPTIONS");
  c.header("Access-Control-Allow-Headers", "Content-Type, Authorization, X-Requested-With");
  c.header("Access-Control-Max-Age", "86400");

  if (c.req.method === "OPTIONS") {
    return c.body(null, 204);
  }

  await next();
}
